File: SpearoGo/Services/GeocodingService.swift
import Foundation

/// Place-name search, for adding a dive spot the diver is not standing at.
///
/// Open-Meteo rather than `CLGeocoder`, even though CoreLocation would work.
/// The two platforms' native geocoders are different services with different
/// results, so the Wear and watchOS apps would disagree about what "Lagos"
/// means. One provider answers both identically and can be pinned in a test
/// fixture. No key, and the same provider already used for weather.
struct GeocodingService {
    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 15
        config.timeoutIntervalForResource = 30
        return URLSession(configuration: config)
    }()

    /// An empty result is a normal answer — "no places found", not a failure.
    func search(_ query: String) async throws -> [GeocodedPlace] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }

        guard var components = URLComponents(string: Constants.API.geocodingBase) else {
            throw ServiceError.invalidURL
        }
        components.queryItems = [
            .init(name: "name", value: trimmed),
            .init(name: "count", value: "8"),
            .init(name: "language", value: "en"),
            .init(name: "format", value: "json"),
        ]
        guard let url = components.url else { throw ServiceError.invalidURL }

        let (data, response) = try await session.data(from: url)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            throw ServiceError.badResponse
        }
        return Self.parse(try JSONDecoder().decode(GeocodingResponse.self, from: data))
    }

    /// Split out so it can be tested against a recorded response.
    static func parse(_ response: GeocodingResponse) -> [GeocodedPlace] {
        (response.results ?? []).compactMap { r in
            guard let lat = r.latitude, let lon = r.longitude,
                  let name = r.name, !name.isEmpty else { return nil }
            // 0,0 is a real place in the Atlantic. A missing coordinate must
            // never become one — the same rule the tide and marine parsing
            // follow.
            return GeocodedPlace(
                id: "\(name)-\(lat)-\(lon)",
                name: name,
                region: r.admin1?.isEmpty == false ? r.admin1 : nil,
                country: r.country_code?.isEmpty == false ? r.country_code : nil,
                latitude: lat,
                longitude: lon
            )
        }
    }
}

struct GeocodingResponse: Decodable {
    let results: [GeocodingResult]?
}

struct GeocodingResult: Decodable {
    let name: String?
    let admin1: String?
    let country_code: String?
    let latitude: Double?
    let longitude: Double?
}

File: wear/app/src/main/java/com/spearotracker/spearogo/services/GeocodingService.kt
package com.spearotracker.spearogo.services

import com.spearotracker.spearogo.models.GeocodedPlace
import com.spearotracker.spearogo.utils.Constants
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import retrofit2.http.Query
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

internal data class GeocodingResponse(val results: List<GeocodingResult>?)

internal data class GeocodingResult(
    val name: String?,
    val admin1: String?,
    val country_code: String?,
    val latitude: Double?,
    val longitude: Double?
)

private interface OpenMeteoGeocodingApi {
    @GET("v1/search")
    suspend fun search(
        @Query("name") name: String,
        @Query("count") count: Int = 8,
        @Query("language") language: String = "en",
        @Query("format") format: String = "json"
    ): GeocodingResponse
}

/**
 * Place-name search, for adding a dive spot the diver is not standing at.
 *
 * Open-Meteo rather than Android's Geocoder, even though Geocoder works on the
 * watch — it is what already turns coordinates into the location label. The two
 * platforms' native geocoders are different services with different results, so
 * the Wear and watchOS apps would disagree about what "Lagos" means. One
 * provider answers both identically and can be pinned in a test fixture. No
 * key, and the same provider already used for weather.
 */
@Singleton
class GeocodingService @Inject constructor() {

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    private val api: OpenMeteoGeocodingApi = Retrofit.Builder()
        .baseUrl(Constants.Api.GEOCODING_BASE)
        .client(client)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(OpenMeteoGeocodingApi::class.java)

    /** Empty is a normal answer — "no places found", not a failure. */
    suspend fun search(query: String): List<GeocodedPlace> {
        val trimmed = query.trim()
        if (trimmed.isEmpty()) return emptyList()
        return parse(api.search(trimmed))
    }

    internal companion object {
        /** Split out so it can be tested against a recorded response. */
        fun parse(response: GeocodingResponse): List<GeocodedPlace> =
            response.results.orEmpty().mapNotNull { r ->
                val lat = r.latitude ?: return@mapNotNull null
                val lon = r.longitude ?: return@mapNotNull null
                val name = r.name?.takeIf { it.isNotBlank() } ?: return@mapNotNull null
                GeocodedPlace(
                    name = name,
                    region = r.admin1?.takeIf { it.isNotBlank() },
                    country = r.country_code?.takeIf { it.isNotBlank() },
                    latitude = lat,
                    longitude = lon
                )
            }
    }
}
